### DISCUSSION TITLE
Open Source Code From "Differentiable Collision Detection: a Randomized Smoothing Approach"?

### DISCUSSION THREAD

**Discussion by alberthli** [stack-of-tasks/pinocchio#1861] [General]
[This recent paper](https://arxiv.org/pdf/2209.09012.pdf) states that the code for differentiable collision detection will be released within the Pinocchio framework. Has this been done already? After a quick search through this repo and the HPP-FCL repo, I couldn't find the implementation. Thank you!

**Comment by jcarpent** [comment]
We are currently working on finishing pinocchio 3, which should also include these collision detection derivatives.

**Reply by joao-pm-santos96** [reply]
This is awesome. Can you provide some high level explanation of how will this be implemented (or the reference to some paper with the math behind it)?
Thank you for your great work!

**Reply by JYChen18** [reply]
Since Pinocchio 3 has already been released, I’ve been trying to locate the code for the derivative of the collision detection but haven’t been successful. Could you provide some guidance on where to find it?

Additionally, I’ve been working on reproducing the results from the _Differentiable Collision Detection_ project on my own but have encountered some challenges. I reached out to your team via email half a month ago but haven’t received a response. Any help would be greatly appreciated!

Thank you!

**Comment by jcarpent** [comment]
@alberthli Which applications do you have in mind of these derivatives?

**Comment by alberthli** [comment]
I see - is there a timeline either for when the implementation of the collision detection derivatives will be rolled into the preview branch or for when pinocchio 3's release is planned?

I was hoping to use the derivatives for robot configuration adjustment based on collisions during planning. In particular, there have been a couple of recent papers on differentiable collision detection, and only pinocchio has Python bindings, which is a necessity to be integrated into my stack.

**Comment by jcarpent** [comment]
> In particular, there have been a couple of recent papers on differentiable collision detection, and only pinocchio has Python bindings, which is a necessity to be integrated into my stack.

From a scientific perspective, the different approaches do not provide the same information.
I don't know precisely which papers you refer to. Still, in our case, we provide derivatives of witness points, which can be used for many usages like distance variation between geometries or contact point variations for simulation.

**Comment by jcarpent** [comment]
> I see - is there a timeline either for when the implementation of the collision detection derivatives will be rolled into the preview branch or for when pinocchio 3's release is planned?

Hard to say: the quickest as we can do, but maintaining large software like Pinocchio and hpp-fcl while also doing research is very demanding ;) Hope you will understand.

**Comment by alberthli** [comment]
> From a scientific perspective, the different approaches do not provide the same information.
I don't know precisely which papers you refer to. Still, in our case, we provide derivatives of witness points, which can be used for many usages like distance variation between geometries or contact point variations for simulation.

Yes, the approaches I've seen all have their differences. Aside from the question of bindings, the witness point approach presented by your collaborators and you is the closest to what I had in mind when thinking about this problem, and would be very useful for my application.

> Hard to say: the quickest as we can do, but maintaining large software like Pinocchio and hpp-fcl while also doing research is very demanding ;) Hope you will understand.

Of course! Thanks for everyone's hard work. It's a great service to the community and I will be watching with interest.

**Comment by sea-bass** [comment]
@jcarpent @lmontaut I just had an intern ask about this and figured I'd check with you guys if you had made any progress on an openly available version of this code? Thanks!

**Reply by jorisv** [reply]
Hello @sea-bass,

Sorry there is no progress right now on this topic.
We still plan to work on it but it's not our priority right now.

**Reply by sea-bass** [reply]
Understood -- Thanks for the update!

**Comment by JYChen18** [comment]
Hi @sea-bass — since you asked about an openly available implementation, I wanted to share a related update.

Following our efforts to reproduce and extend differentiable collision detection based on randomized smoothing, we developed an independent follow-up work, [Robust Differentiable Collision Detection for General Objects](https://arxiv.org/abs/2511.06267) (ICRA 2026), and released the accompanying PyTorch implementation: [JYChen18/DiffCollision](https://github.com/JYChen18/DiffCollision)

DiffCollision builds on Coal for forward collision queries and provides a PyTorch interface with autograd support. It includes several differentiable backends and supports both convex and concave geometries. For clarity, this is not the original Pinocchio implementation discussed here, nor is it an official Pinocchio or Coal component. It is an independent follow-up project.

We are continuing to explore integration with differentiable simulation, GPU acceleration, and improvements over the current formulation. These directions are still experimental, but we hope the released repository can already provide a useful starting point for others interested in this problem. Feedback and discussion are very welcome!
